### PR TITLE
OverDrive: Support disabled patron API

### PR DESCRIPTION
**Pull request by maccabeelevine**
The Overdrive configuration instructions say to request patron authentication access, which is needed for the various patron functionality in the integration (checkouts, holds, etc.).  However patron authentication is not supported by Overdrive if your institution uses SAML authentication for the Overdrive UI.  I got this in an email from OD:

> Looking at your current authentication settings, I see that your users are authenticating via a SAML connection. Unfortunately, our circulation APIs do not support external authentication methods. It is possible that your catalog/OPAC vendor has an integration option, to only utilize the Discovery APIs (and forward users to the OverDrive site for circulation); however, you would need to discuss the integrations available options with your vendor. ...

In theory we could disable SAML authentication, if the goal is to always interact via the VuFind UI.  But I think it's a good idea to support SAML where it's possible to do so. 

So this PR allows the VuFind integration to optionally work at the the discovery API level only.  It still displays buttons for patron actions, but they link to the OD UI record page.  This also avoids two OD API calls each time a record page is displayed that will always fail and will generate error messages.

### CODE REVIEW

**Comment by maccabeelevine** [comment]
@bpalme hoping for your input on this!

File: themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml
Comment (maccabeelevine) [inline themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:61]: I see that there is a scenario where `url` is absent and `route` is there instead, but I'm not sure there is a way for that to happen and for `supportsPatronActions` to also be false.  That said I would like to protected against the possibility, but I'm not sure what to do if it happened.  I suppose I could just have the check at line 55 be more complicated, i.e. `$supportsPatronActions && ($urls[0]['url'] ?? false)`, but  but there might be a better idea.

This affects line 74 as well.

File: module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php
Comment (maccabeelevine) [inline module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php]: This code will presumably never be reached since the UI would now not present options to do a checkout; same with most of the other mutation functions below.  But I think it's better to catch the scenario just in case.

File: config/vufind/Overdrive.ini
Comment (demiankatz) [inline config/vufind/Overdrive.ini]: Typo: authentiation.

Comment (demiankatz) [reply config/vufind/Overdrive.ini]: Also, I think the preferred capitalization is OverDrive, so you may want to do a string-replace of Overdrive --> OverDrive for consistency.

File: module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php]: Is this a stray comment that wasn't meant to be committed?

**Review by demiankatz — CHANGES_REQUESTED** [review]
I defer to @bpalme for commentary on the actual functional changes, but I found a couple of minor nitpicks to report in the meantime. ;-)

File: config/vufind/Overdrive.ini
Comment (maccabeelevine) [reply config/vufind/Overdrive.ini]: Thanks, I fixed the capitalization here in Overdrive.ini.  I could fix the new log messages in OverdriveConnector, but then I'd probably also want to fix the *old* log messages?  (I would not touch the filenames / function names, we don't want another "fulfilment" breaking change...)

File: module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php
Comment (maccabeelevine) [reply module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php]: Yup, fixed.

File: config/vufind/Overdrive.ini
Comment (demiankatz) [reply config/vufind/Overdrive.ini]: Agreed, I'm not too worried about file or function names, but I'm trying to be consistent in things that are more user-facing like on-screen messages and documentation. It wouldn't hurt to fix the log messages too, though it's not absolutely critical. If you feel like doing it, please go ahead, but it might also be something we could tackle as a follow-up PR if we want to keep this one from getting too many unrelated changes.

Comment (maccabeelevine) [reply config/vufind/Overdrive.ini]: Ok I think I took care of all the non-breaking capitalization in OverdriveConnector, where it makes sense to do now since there are so many log messages already and we are adding more.

Comment (demiankatz) [reply config/vufind/Overdrive.ini]: Looks good to me -- thanks!

**Comment by bpalme** [comment]
I just got back from a trip. I plan to take a look at this in the next couple of days.

**Comment by bpalme** [comment]
I did a little testing with it on and off in my environment and it looks good to me. Good configuration option I think.

File: themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml
Comment (demiankatz) [reply themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:61]: You could support route easily enough, with something like:

`$urls[0]['url'] ?? $this->url($urls[0]['route'] ?? 'home', $urls[0]['routeParams'] ?? [])`

But I don't think we expect a route here, so that's probably going to lead the user to a weird and unexpected place.

Another option might be to leverage the permission denied error page, like this:

`$urls[0]['url'] ?? $this->url('error-permissiondenied', [], ['query' => ['msg' => 'od_docheckout_failure']])`

That should take the user to this:

image

...which isn't _exactly_ accurate, but is probably close enough to be helpful (by indicating the checkout failed and linking to the support email) if this suspected-impossible situation were to actually arise.

**Review by demiankatz — APPROVED** [review]
I'm approving this since @bpalme approves... but I also responded to the issue that @maccabeelevine raised with a possible solution. I will hold off on merging this until he has a chance to address that (if he wishes to).

File: themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml
Comment (maccabeelevine) [reply themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:61]: I gave the `error-permissiondenied` solution a try and forced that error state.  However I forgot the whole button has a data-lightbox attribute, so what actually happens is this, which I think is unhelpful.

image

I could fix that with a further check on the presence of `$urls[0]['url']` before adding the lightbox attribute, but I think this has gone to the point where the code becomes confusing and implies valid states that should not actually exist.  So I think actually no change is better, and apologies in advance if this error actually comes up -- would need a smarter solution then based on why it actually happens.

Comment (demiankatz) [reply themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:61]: Drat, sorry that I didn't think of the lightbox situation. Might it be better to just throw an exception (or, as a softer approach, log an error) if an unexpected state occurs, so it can be tracked down if it does? Or should we just hope for the best? :-)

Comment (maccabeelevine) [reply themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:61]: I decided to take a different approach here and consider the missing URL not to be an error in the first place, by just changing the `count($urls)` check to actually verify that $urls[0] has a URL.  So if there is no URL, there will just be no button, the same as if $urls was empty -- no more or less valid.  If this every comes up in reality, it could be for a good reason.

Comment (demiankatz) [reply themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:61]: That's a great solution; thank you!

**Comment by demiankatz** [comment]
Looks like this is ready -- merging now. Thanks again!

**Pull request merged**
